File: clients/python/smoke.py
"""Browser-free Python client handshake smoke test."""

import os
import json
from pathlib import Path

from glass_client import GlassClient, GlassError


binary = os.environ.get(
    "GLASS_BINARY",
    str(Path(__file__).resolve().parents[2] / "target" / "debug" / "glass"),
)
client = GlassClient(command=binary)
try:
    manifest = client.initialize()
    assert manifest["protocolVersion"] == 1
    assert client.supports_capability("action")
    browser_fixture = json.loads(
        (
            Path(__file__).resolve().parents[2]
            / "crates"
            / "glass-browser"
            / "tests"
            / "fixtures"
            / "client-conformance-v1.json"
        ).read_text()
    )
    development_fixture = json.loads(
        (
            Path(__file__).resolve().parents[2]
            / "crates"
            / "glass-dev"
            / "tests"
            / "fixtures"
            / "client-conformance-v1.json"
        ).read_text()
    )
    for schema, versions in browser_fixture["requiredSchemas"].items():
        for version in versions:
            assert client.supports_schema(schema, version)
    for capability in browser_fixture["requiredCapabilities"]:
        client.require_capability(capability)
    tool_names = sorted(tool["name"] for tool in client.list_tools())
    assert tool_names == development_fixture["tools"]
    trust = client.call("glass.workspace.trust.status")
    assert trust["trust"] in {"untrusted", "trustedOnce", "trustedProject"}
    authority = client.call("glass.workspace.trust.inspect")
    assert isinstance(authority["items"], list)
    tree = client.call("glass.file.list")
    assert isinstance(tree["entries"], list)
    browser = client.call("glass.browser.state")
    assert isinstance(browser["connected"], bool)
    assert isinstance(client.call("glass.task.list"), list)
    assert isinstance(client.call("glass.replay.list"), list)
    if trust["trust"] == "untrusted":
        try:
            client.call("glass.test.discover")
        except GlassError as error:
            assert "trusted" in str(error).lower() or "blocked" in str(error).lower()
        else:
            raise AssertionError("untrusted executable project discovery was not denied")
finally:
    client.close()

File: clients/typescript/smoke.mjs
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { GlassClient, GlassStructuredError } from "./dist/index.js";

const repositoryRoot = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "../..");

const client = new GlassClient({
  command: process.env.GLASS_BINARY ?? path.join(repositoryRoot, "target", "debug", "glass"),
});
try {
  const manifest = await client.initialize();
  if (manifest?.protocolVersion !== 1) throw new Error("unexpected Glass protocol version");
  if (!client.supportsCapability("action")) throw new Error("action capability missing");
  const browserFixture = JSON.parse(fs.readFileSync(path.join(repositoryRoot, "crates/glass-browser/tests/fixtures/client-conformance-v1.json"), "utf8"));
  const developmentFixture = JSON.parse(fs.readFileSync(path.join(repositoryRoot, "crates/glass-dev/tests/fixtures/client-conformance-v1.json"), "utf8"));
  for (const [schema, versions] of Object.entries(browserFixture.requiredSchemas)) {
    for (const version of versions) if (!client.supportsSchema(schema, version)) throw new Error(`schema missing: ${schema}@${version}`);
  }
  for (const capability of browserFixture.requiredCapabilities) client.requireCapability(capability);
  const toolNames = (await client.listTools()).map((tool) => tool.name).sort();
  if (JSON.stringify(toolNames) !== JSON.stringify(developmentFixture.tools)) throw new Error("MCP tool inventory does not match the full-product conformance fixture");
  const trust = await client.call("glass.workspace.trust.status");
  if (!trust || typeof trust !== "object" || !["untrusted", "trustedOnce", "trustedProject"].includes(trust.trust)) {
    throw new Error("invalid workspace trust status");
  }
  const authority = await client.call("glass.workspace.trust.inspect");
  if (!authority || typeof authority !== "object" || !Array.isArray(authority.items)) {
    throw new Error("invalid workspace authority inspection");
  }
  const tree = await client.call("glass.file.list");
  if (!tree || typeof tree !== "object" || !Array.isArray(tree.entries)) {
    throw new Error("invalid bounded file list");
  }
  const browser = await client.call("glass.browser.state");
  if (!browser || typeof browser !== "object" || typeof browser.connected !== "boolean") {
    throw new Error("invalid resident browser state");
  }
  if (!Array.isArray(await client.call("glass.task.list"))) throw new Error("task list was not an array");
  if (!Array.isArray(await client.call("glass.replay.list"))) throw new Error("replay list was not an array");
  if (trust.trust === "untrusted") {
    try {
      await client.call("glass.test.discover");
      throw new Error("untrusted executable project discovery was not denied");
    } catch (error) {
      if (!(error instanceof GlassStructuredError)) throw error;
    }
  }
} finally {
  client.close();
}
